docs/design/ha_master.md

```diff
@@ -0,0 +1,255 @@
+will be forwarded by the local etcd server such that it’s invisible for the user). It will
+affect latency for all operations, but it should not increase by much more than the network
+latency between master replicas (latency between GCE zones with a region is < 10ms).
+
+Currently etcd exposes port only using localhost interface. In order to allow clustering
+and inter-VM communication we will also have to use public interface. To secure the
+communication we will configure firewall rules to only allow traffic on etcd port from
+other master machines.
+
+Another option to secure etcd cluster communication would be to use SSL (as described
+[here](https://coreos.com/etcd/docs/latest/security.html)), but it would increase CPU usage
+on master machines, which is already a bottleneck in large clusters. In many
+use cases it is not required so we will not offer this in the first version.
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all existing kubernetes master replicas (see `VM naming in GCE` section).
+Based on that, we will set the following flags:
+* `-initial-cluster` - list of all hostnames/DNS names for master replicas (including the new one)
+* `-initial-cluster-state` (keep in mind that we are adding master replicas one by one):
+  * `new` if we are adding the first replica, i.e. the list of existing master replicas is empty
+  * `existing` if there are more than one replica, i.e. the list of existing master replicas is non-empty.
+
+This will allow us to have exactly the same logic for HA and non-HA master. List of DNS names for VMs
+with master replicas will be generated in `kube-up.sh` script and passed to as a env variable
+`INITIAL_ETCD_CLUSTER`.
+
+### apiservers
+
+All apiservers will work independently. They will contact etcd on 127.0.0.1, i.e. they will always contact
+etcd replica running on the same VM. If needed, such requests will be forwarded by etcd server to the
+etcd leader. This functionality is completely hidden from the client (apiserver
+in our case).
+
+Caching mechanism, which is implemented in apiserver, will not be affected by
+replicating master because:
+* GET requests go directly to etcd
+* LIST requests go either directly to etcd or to cache populated via watch
+  (depending on the ResourceVersion in ListOptions). In the second scenario,
+  after a PUT/POST request, changes might not be visible in LIST response.
+  This is however not worse than it is with the current single master.
+* WATCH does not give any guarantees when change will be delivered.
+
+#### DNS & load balancer
+
+In order to expose a single endpoint for the user, we will introduce a public DNS name for the cluster.
+For a non-replicated master it will point directly to the apiserver. For a replicated master
+we will have two options:
+
+* create an L4 load balancer in front of all apiservers and update DNS name appropriately
```

Comment:
What are we trying to do with this doc? Are we describing a general spec for all automated HA cluster deployments or are we describing what we are actually going to do in the next 6 months? 

> For example in GCP it's easier to configure L4 than L7 load balancer.

Given it's current feature set, it's not actually possible to front an api server with an L7 load balancer.